src/HexagonOptimize.cpp

```diff
@@ -0,0 +1,1084 @@
+
+struct Pattern {
+    enum Flags {
+        InterleaveResult = 1 << 0,  ///< After evaluating the pattern, interleave native vectors of the result.
+        SwapOps01 = 1 << 1,  ///< Swap operands 0 and 1 prior to substitution.
+        SwapOps12 = 1 << 2,  ///< Swap operands 1 and 2 prior to substitution.
+        ExactLog2Op1 = 1 << 3, ///< Replace operand 1 with its log base 2, if the log base 2 is exact.
+        ExactLog2Op2 = 1 << 4, ///< Save as above, but for operand 2.
+
+        DeinterleaveOp0 = 1 << 5,  ///< Prior to evaluating the pattern, deinterleave native vectors of operand 0.
+        DeinterleaveOp1 = 1 << 6,  ///< Same as above, but for operand 1.
+        DeinterleaveOp2 = 1 << 7,
+        DeinterleaveOps = DeinterleaveOp0 | DeinterleaveOp1 | DeinterleaveOp2,
+
+        NarrowOp0 = 1 << 10,  ///< Replace operand 0 with its half-width equivalent.
+        NarrowOp1 = 1 << 11,  ///< Same as above, but for operand 1.
+        NarrowOp2 = 1 << 12,
+        NarrowOps = NarrowOp0 | NarrowOp1 | NarrowOp2,
+
+        NarrowUnsignedOp0 = 1 << 15,  ///< Similar to the above, but narrow to an unsigned half width type.
+        NarrowUnsignedOp1 = 1 << 16,
+        NarrowUnsignedOp2 = 1 << 17,
+        NarrowUnsignedOps = NarrowUnsignedOp0 | NarrowUnsignedOp1 | NarrowUnsignedOp2,
+    };
+    string intrin;        ///< Name of the intrinsic
+    Expr pattern;         ///< The pattern to match against
+    int flags;
+
+    Pattern() {}
+    Pattern(const string &intrin, Expr p, int flags = 0)
+        : intrin(intrin), pattern(p), flags(flags) {}
+};
+
+Expr wild_u8 = Variable::make(UInt(8), "*");
+Expr wild_u16 = Variable::make(UInt(16), "*");
+Expr wild_u32 = Variable::make(UInt(32), "*");
+Expr wild_u64 = Variable::make(UInt(64), "*");
+Expr wild_i8 = Variable::make(Int(8), "*");
+Expr wild_i16 = Variable::make(Int(16), "*");
+Expr wild_i32 = Variable::make(Int(32), "*");
+Expr wild_i64 = Variable::make(Int(64), "*");
+
+Expr wild_u8x = Variable::make(Type(Type::UInt, 8, 0), "*");
+Expr wild_u16x = Variable::make(Type(Type::UInt, 16, 0), "*");
+Expr wild_u32x = Variable::make(Type(Type::UInt, 32, 0), "*");
+Expr wild_u64x = Variable::make(Type(Type::UInt, 64, 0), "*");
+Expr wild_i8x = Variable::make(Type(Type::Int, 8, 0), "*");
+Expr wild_i16x = Variable::make(Type(Type::Int, 16, 0), "*");
+Expr wild_i32x = Variable::make(Type(Type::Int, 32, 0), "*");
+Expr wild_i64x = Variable::make(Type(Type::Int, 64, 0), "*");
+
+std::vector<Pattern> casts = {
+    // Averaging
+    { "halide.hexagon.avg.vub.vub", u8((wild_u16x + wild_u16x)/2), Pattern::NarrowOps },
+    { "halide.hexagon.avg.vuh.vuh", u16((wild_u32x + wild_u32x)/2), Pattern::NarrowOps },
+    { "halide.hexagon.avg.vh.vh", i16((wild_i32x + wild_i32x)/2), Pattern::NarrowOps },
+    { "halide.hexagon.avg.vw.vw", i32((wild_i64x + wild_i64x)/2), Pattern::NarrowOps },
+
+    { "halide.hexagon.avg_rnd.vub.vub", u8((wild_u16x + wild_u16x + 1)/2), Pattern::NarrowOps },
```

Comment:
The simplifier ensures that the constant is outermost and last. Perhaps the arm backend is being too cautious. Or maybe expressions are constructed during codegen that never make it through the simplifier.